workflows/pr: preparation for required status checks

wolfgangwalther opened the pull request:
This adds a `PR / required to merge` job, which:
- is skipped when Check / Lint / Eval / Build all succeed,
- fails when any of them fail.

Once merged, this job can be used for the "Required Status Checks" branch protection rule. Once we enable that, merging PRs will only be possible when CI succeeds.

Before adding this job, I restructed the PR workflow slightly: Labels and Reviewers were previously part of Eval, but that would make them required as well. By moving them up into the PR workflow, they can be optional for the `required` job. They don't really need to block a merge.

While doing that, I also fixed the merge commit inconsistencies described in https://github.com/NixOS/nixpkgs/pull/406825#issuecomment-2888568813.

Roadmap to Required Status Checks:
- [x] Review, approve, merge this PR
- [ ] Optional: Wait a while, so that the required job has made its way into a few more PRs.
- [ ] Enable the branch protection rule. See https://github.com/NixOS/org/issues/130

I'm not sure if we need to wait and if yes, how long, before enabling the checks. Every PR that ran its CI the last time before the merge of this PR, will be blocked from merging until CI is triggered again.

Note: Any discussion around *if and how to implement the branch protection rules* is better placed in https://github.com/NixOS/org/issues/130.

## Things done

- [x] Tested in my fork.
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
workflows/{pr,push}: move prepare job from eval workflow

This is only a refactor at this stage, but split into a separate commit
for better review. It's the base for the next two commits.

wolfgangwalther committed:
workflows: sync merge commits

This fixes a problem where each workflow would get their own merge
commit. This happens frequently when the target branch is merged into a
the same time, different workflows in the same run will run
get-merge-commit at different times and thus have different merge
commits.

Since the jobs don't really depend on each other, this doesn't cause
practical problems, yet. But it has already led to strange CI failures
in a still unmerged PR, which can be prevented from happening with this
clean approach.

And yes, this saves a few API calls on every run.

wolfgangwalther committed:
workflows/{labels,reviewers}: move from Eval to PR context

This allows *not* depending on those two jobs with the required status
checks in the next commit, which wouldn't really make sense. If labeling
or pinging maintainers fails for obscure reasons or because the GitHub
API is down, a PR might still pass all other tests and be
merge-eligible.

wolfgangwalther committed:
workflows/pr: add required job

This job serves as a target for the "Required Status Checks" branch
protection rule.

MattSturgeon commented on .github/workflows/pr.yml:
```diff
@@ -27,21 +48,68 @@ jobs:
+      - build
+    # WARNING: Do NOT change the name of this job, otherwise branch protection rules
+    # will not catch it anymore. This will prevent all Pull Requests from merging.
+    name: required to merge
```
[In my personal config](https://github.com/MattSturgeon/nix-config/blob/4bf1f4e4a13b86a08c6551d8c89529eb48695f5c/.github/workflows/check.yml#L76) I named it "success", to indicate that this would happen at the end of the workflow run.

That'd show up as "1 pending check: PR / success"
Vs the current: "1 pending check: PR / required to merge"

Another option: "1 pending check: PR / ok"
Another option: "1 pending check: PR / passes"
Another option: "1 pending check: PR / no issues"
Another option: "1 pending check: PR / no failures"

Naming this meta job isn't easy; I'm not sure _any_ name really makes sense for a job that is required to be skipped :grin: 

But worth getting more nit-picky over the name now, _before_ we hardcode it into our ruleset!

philiptaron replied:
`needed PR checks complete for nixpkgs CI`?

philiptaron replied:
`gate for merge` looks the best to me.

MattSturgeon commented on .github/workflows/pr.yml:
```diff
@@ -27,21 +48,68 @@ jobs:
+  # However, if any of the previous jobs fails, this will run and fail itself as well - and thus
+  # prevent from merging the Pull Request.
+  required:
+    # If you need to add or remove jobs from the list of required status checks, modify this list.
```
```suggestion
    # Modify this list to add or remove jobs from required status checks.
```

or

```suggestion
    # This list defines which jobs are required status checks.
```

or

```suggestion
    # Jobs that are required status checks:
```

or

   
```suggestion
    # The PR cannot merge if any of these fail:
```

or simply:
   
```suggestion
    # Modify this list as required:
```

MattSturgeon commented on .github/workflows/pr.yml:
```diff
@@ -27,21 +48,68 @@ jobs:
+    with:
+      mergedSha: ${{ needs.prepare.outputs.mergedSha }}
+
+  # This job's only purpose is to serve as a target for the "Required Status Checks" branch
+  # protection rule. It depends on all the jobs that are required to pass. If all tests pass, it
+  # doesn't run itself, because "skipped by conditional" counts as "success" in terms of those
+  # branch protection rules.
+  # However, if any of the previous jobs fails, this will run and fail itself as well - and thus
+  # prevent from merging the Pull Request.
```
Editing pass on this paragraph:

```suggestion
  # This job's only purpose is to serve as a target for the "Required Status Checks" ruleset.
  # It "needs" all the jobs that should block merging a PR. If they pass, it is skipped —
  # which counts as "success" for purposes of the ruleset.
  # However, if any of them fail, this job will also fail — thus blocking the ruleset.
```

MattSturgeon commented on .github/workflows/pr.yml:
```diff
@@ -27,21 +48,68 @@ jobs:
+      - lint
+      - eval
+      - build
+    # WARNING: Do NOT change the name of this job, otherwise branch protection rules
+    # will not catch it anymore. This will prevent all Pull Requests from merging.
```
I just had a quick look, and I was somewhat misremembering the UI. I'll hide my last comment as "outdated".

Even for checks that exist in a repo, once added they are just saved as the job name. You do get a search box that "shows" the actual job, but once saved you only have the nameThe only other metadata saved is the "integration_id", which in this case indicates that it is GitHub Actions, rather than some other App like buildbot or ofborg.

E.g. my nix-config rule looks like:
![image](https://github.com/user-attachments/assets/d3e89d93-27ff-409e-9b42-efbe8909043b)

and is exported as:
```json
"required_status_checks": [
  {
    "context": "Success",
    "integration_id": 15368
  }
]
```

For reference, the workflow job is defined as:
```yaml
check-success:
  name: Success
```

So yes, there is definitely value in the job name being somewhat unique.

wolfgangwalther committed:
workflows/eval: small improvements

Some naming improvements after we introduced the PR / Push workflows and
small refactors.

MattSturgeon commented on .github/workflows/push.yml:
```diff
@@ -18,12 +18,32 @@ on:
 permissions: {}
 
 jobs:
+  prepare:
+    runs-on: ubuntu-24.04-arm
+    outputs:
+      systems: ${{ steps.systems.outputs.systems }}
+    steps:
+      - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
+        with:
+          sparse-checkout: |
+            ci/supportedSystems.json
+
+      - name: Load supported systems
+        id: systems
+        run: |
+          echo "systems=$(jq -c <ci/supportedSystems.json)" >> "$GITHUB_OUTPUT"
```
> The problem with _local_ actions is... they need a _checkout_ before!

In the other usage, you're already checking out `.github/actions`, so that could be common to both. But I assume what you're saying is that _here_ you'd be checking out a potentially untrusted `.github/actions` revision?

MattSturgeon commented on .github/workflows/pr.yml:
```diff
@@ -27,21 +48,67 @@ jobs:
+    with:
+      mergedSha: ${{ needs.prepare.outputs.mergedSha }}
+
+  # This job's only purpose is to serve as a target for the "Required Status Checks" branch ruleset.
+  # It "needs" all the jobs that should block merging a PR.
+  # If they pass, it is skipped — which counts as "success" for purposes of the branch ruleset.
+  # However, if any of them fail, this job will also fail — thus blocking the branch ruleset.
+  no-pr-failures:
+    # Modify this list to add or remove jobs from required status checks.
+    needs:
+      - check
+      - lint
+      - eval
+      - build
+    # WARNING:
+    # Do NOT change the name of this job, otherwise the rule will not catch it anymore.
+    # This would prevent all PRs from merging.
+    name: no PR failures
+    if: ${{ failure() }}
+    runs-on: ubuntu-24.04-arm
+    steps:
+      - run: exit 1
```
I assume, before merging, we should confirm in https://github.com/NixOS/org/issues/130 that we are definitely planning to enable required status checks at some point?

I can't see why we _wouldn't_ want to... But if the proposal was blocked for any reason, then we wouldn't need to merge this job.

---

Alternatively, we could merge it optimistically so that we can see it working in production.

Once we've proven the CI is working, it will be easier to argue in favour of requiring it.

With this approach, we can still remove the job later if the proposal is blocked or we change our minds (for whatever reason).

MattSturgeon reviewed (approved):
Went through each commit again. Everything seems clear, and I don't see any obvious issues.

I'd say let's test this out :rocket:
